In [17]:
import pandas as pd
import numpy as np
import random
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

pd.set_option('display.max_columns', None)

## Load the original Dataset

In [18]:
df = pd.read_csv("loan_continuous.csv")

### Check Loan Status Column and lenght of dataset

In [19]:
df['Loan Status'].value_counts()

0    61222
1     6241
Name: Loan Status, dtype: int64

In [20]:
len(df)

67463

## Load the expanded Dataset

In [21]:
df_e = pd.read_csv("loan_continuous_expanded.csv")

### Check Loan Status Column and lenght of dataset

In [22]:
df_e['Loan Status'].value_counts()

0    72598
1    44865
Name: Loan Status, dtype: int64

In [23]:
len(df_e)

117463

## Load fake Dataset

In [24]:
fake = pd.read_csv("fake.csv")

## Visually compare samples of:
- Original Data
- Subset of Loan Status Class 1 data
- Fake data

In [25]:
# Original data sample
df.sample(20)

Loan Amount  Funded Amount  Funded Amount Investor  Term  \
6454          5790           9302            23724.130070    59   
62319        18059          31301             9927.005719    59   
38357        13075          11240            13464.514980    59   
52416        12253          33132            32370.326010    58   
62720         8692           3504             8994.612305    59   
12324        17412           7260             4902.637738    58   
50716        25668          30402            10702.099220    59   
48841        14395          11319            23128.454790    59   
9191         22365           9695            17117.931720    59   
19820         4369          12542            27232.455880    59   
43773        17513          14825            15724.191100    59   
7792         10198          32336            11029.898460    59   
50412        22582          31946            11542.339410    59   
45331        15561          31766            16041.430730    59   
43198        21794          19868            17146.868620    58   
28415         3923           9383            10149.115480    58   
4178          2964           2742            25493.766640    59   
23859        27328           7821             8169.950946    58   
65490        13446           6980             8611.584326    59   
65962        15179          11304            29632.119110    58   

       Home Ownership  Debit to Income  Delinquency - two years  \
6454      32452.69646        37.469399                        0   
62319    140619.88050        30.766134                        1   
38357     69975.77784        19.480475                        1   
52416     49401.77311        24.847906                        1   
62720     60628.14107        15.732289                        0   
12324     92289.95366        33.187576                        0   
50716    126735.55370        14.150859                        0   
48841    140398.91080        36.042365                        0   
9191     111924.47680        19.134003                        0   
19820     71729.61011        22.888129                        0   
43773    111914.40800        24.749360                        1   
7792     138882.95590        13.820506                        0   
50412     60957.88255        26.220642                        0   
45331     76887.24647        18.641615                        0   
43198     75230.55668        17.383202                        0   
28415     59121.00810        23.387476                        0   
4178      37219.53547        37.534723                        0   
23859     62576.68351        16.416342                        0   
65490    133148.52130        27.948796                        1   
65962     45924.05152        12.090613                        0   

       Inquires - six months  Open Account  Public Record  Revolving Balance  \
6454                       0            13              0               4711   
62319                      0            13              0                593   
38357                      0            12              0              11471   
52416                      0            10              0                661   
62720                      2             7              0               3990   
12324                      0             6              0               2024   
50716                      0             7              0               3551   
48841                      0            12              0               4173   
9191                       0            11              0               4057   
19820                      0            16              0              19250   
43773                      0             7              0               7986   
7792                       0            11              0              19606   
50412                      3            15              0              15323   
45331                      0            22              0              

In [13]:
# Sample of Subset of Data with Loan status == 1
df[df['Loan Status'] == 1].sample(20)

Loan Amount  Funded Amount  Funded Amount Investor  Term  \
63126        16775           5353            14332.102840    58   
44267        26126          11683            14060.242990    58   
55233        33943          28900            24049.472750    59   
30650        22264          21854             8001.344148    59   
65910         5008          11908            21203.529510    58   
27945         5118          22480            15261.214450    59   
4160         16919          13347             9715.956808    58   
44389        14900           5269            14529.060200    59   
31114         9709           3513             9711.341237    59   
5326         23025           5780            11750.635300    59   
38951         3271          11193            10418.901510    59   
28790        10904          21508            24691.725880    58   
55640        21704           9908            11070.139180    59   
53277        12556           6710            14210.073840    36   
32308        16251          30610             9048.376524    58   
6968         13776          13688            13483.425390    59   
16850        15847           6730             7871.235250    59   
13276        31430          15485            11261.330970    58   
11252        31615           9158            23371.855920    58   
2635         15534          26629            10138.696020    59   

       Home Ownership  Debit to Income  Delinquency - two years  \
63126     94150.80853        22.783107                        0   
44267     93012.34678        18.864395                        0   
55233     63261.40341        27.554252                        0   
30650    111039.79440        23.654801                        0   
65910    368216.25920        20.699472                        0   
27945    104449.39940        24.364751                        0   
4160      51723.35010        20.090509                        1   
44389     92890.98624        11.987553                        0   
31114     71223.49404         9.828915                        0   
5326     148069.13350        18.663331                        0   
38951     69748.96405        39.360980                        3   
28790    135399.17830        20.248103                        3   
55640     40193.02107        23.742753                        1   
53277    133218.18390        17.741337                        0   
32308     60145.57149        38.420467                        0   
6968      60743.84385        30.753675                        0   
16850    217869.55310         9.675055                        1   
13276     84211.12748         8.467685                        0   
11252     56921.10182        35.434319                        0   
2635      64797.35505        28.466195                        0   

       Inquires - six months  Open Account  Public Record  ...  \
63126                      0            18              0  ...   
44267                      0            19              0  ...   
55233                      0            10              1  ...   
30650                      0             8              0  ...   
65910                      0            12              0  ...   
27945                      0            14              0  ...   
4160                       0            29              0  ...   
44389                      0            27              0  ...   
31114                      0            24              0  ...   
5326                       0            13              0  ...   
38951                      0            15              0  ...   
28790                      0            11              0  ...   
55640                      0            21              0  ...   
53277                      0            32              0  ...   
32308                      0            28              0  ...   
6968                       1            16              0  ...   
16850                      0            27              1  ...   
132

In [26]:
# Sample of generated synthetic data
fake.sample(30)

Loan Amount  Funded Amount  Funded Amount Investor       Term  \
14687   15900.8230     17478.5250             17830.01800  57.100254   
22248   20848.2730     29567.2850             21842.06000  61.624760   
13735   11195.7420     10156.7160             14578.01600  54.467766   
2422    28710.4080     14786.8080             14456.21800  61.746864   
11579   20487.5490      9619.7110             14671.00400  58.897774   
15392   11565.2320     23459.1130             13855.56200  57.331074   
16207    9017.8230     14207.2540             12449.77900  54.045315   
45117   13017.5630      6440.2300             12999.98700  59.484220   
12083   28496.8610     14539.7850             14541.98900  61.656487   
22021   29689.5550     33847.0100             29752.98400  64.007850   
34701   12407.0630     26187.5250             12655.46300  57.828896   
26196   18202.6950     27732.3930             17615.52000  59.923355   
3959     4055.0718      7197.3580             27052.37300  57.140987   
17092   15378.7590      6682.6655             12210.78800  56.280870   
8642    15846.6980      9652.8030             13659.81200  56.313900   
41463   26382.6520     13573.6780             16463.01400  61.105103   
10971   23151.6290     12952.0050             18024.67600  60.350624   
25842   14554.7420      8090.4897              8905.07500  56.705517   
1745    11607.5580     11054.3790              5953.74660  57.361126   
35712   11565.7500     23451.4900             13860.83800  57.334194   
19669   30903.5080     34258.6130             30797.77000  64.291240   
40044   18525.8380     14252.1160               898.57263  48.229370   
1840    19401.7750     28487.8180             19522.11700  60.728832   
42991    5914.1890      8226.1610             20421.92600  57.898994   
47038   13384.0130      5921.4966             15162.86100  58.862175   
41113   13870.0300     20103.3520             16106.82800  57.240060   
40285   17549.1230     13422.8790              4984.11200  49.982903   
25746   12197.1730     26077.2540             12520.52050  57.601254   
20148   17573.3600     14368.3370             18823.85500  58.259880   
35326   11494.0170     11995.3860             14102.94800  58.318604   

       Home Ownership  Debit to Income  Delinquency - two years  \
14687      102479.300        29.686682                -0.218333   
22248       79954.880        29.243834                -0.718972   
13735      102989.380        23.274143                 0.174903   
2422        21899.287        13.594836                 0.548471   
11579       87712.640        17.395310                 0.695672   
15392      112442.760        29.617882                -0.733247   
16207      115691.875        24.335710                 0.077271   
45117       81594.560        15.015101                 0.851209   
12083       25905.947        13.606773                 0.529346   
22021       44182.180        34.384747                -0.918099   
34701      112201.640        25.064352                -0.626088   
26196       94220.020        26.483437                -0.666892   
3959       122705.020        33.533924                 0.934423   
17092      104333.760        18.166842                 0.858503   
8642       105505.040        18.257229                 1.106446   
41463       54460.720        16.677547                 0.301524   
10971       62344.434        19.981003                 0.297890   
25842      118268.695        17.525084                 1.161856   
1745        97034.400        18.121428                 1.729437   
35712      112410.445        29.621965                -0.733477   
19669       38013.242        35.288790                -0.937045   
40044      119748.125        20.654928                 0.159790   
1840        87349.110        27.736790                -0.688505   
42991       99525.220        28.700302                 0.685252   
47038       94236.000        18.383780                 0.693357   
41113      110941.530        31.

### Inspecting generated synthetic data

In [27]:
# Threshold was set to 0.5 for float to categorical conversion of the Loan Status Column
len(fake[fake['Loan Status'] >= 0.5])

38624

In [12]:
# Percentage generated Loan Status == 1 rows
fake['Loan Status'].mean()

0.77248

**Observation:** Expected was a generation of data which columns are only **floats**, and therefore this data differs from the imbalanced part of the original dataset. The set threshold of 0.5 for categorical conversion of the target column **Loan Status**, led to 77% Loan Status == 1 rows. Shifting this threshold would result in different percentages and influences the balancing attempt of the dataset.

Further the synthetic data could/should be made more realistic by manually converting other features back to categorical data, amongst other measures that can be taken to try to optimise the process.